src/coordinator/coordinator_server.cc

```diff
@@ -0,0 +1,441 @@
+  cohort_request.set_transaction_id(transaction_id);
+  std::vector<std::unique_ptr<ClientContext>> transaction_result_contexts;
+  absl::flat_hash_map<std::string,
+                      std::unique_ptr<grpc::ClientAsyncResponseReaderInterface<
+                          cohort::GetTransactionResultResponse>>>
+      async_responses_by_namespace;
+  size_t num_committed_responses = 0;
+  for (const auto &ns :
+       metadata_by_transaction_[transaction_id].get_cohort_namespaces) {
+    if (metadata_by_transaction_[transaction_id]
+            .get_cohorts_already_responded.contains(ns.address())) {
+      ++num_committed_responses;
+    } else {
+      transaction_result_contexts.push_back(
+          ClientContext::FromServerContext(context));
+      async_responses_by_namespace[ns.address()] = AsyncGetResultsFromCohort(
+          ns, cohort_request, *transaction_result_contexts.back());
+    }
+  }
+  for (auto &namespace_and_response : async_responses_by_namespace) {
+    cohort::GetTransactionResultResponse cohort_response;
+    grpc::Status status = FinishAsyncGetResultsFromCohort(
+        namespace_and_response.second, cohort_response);
+    if (status.ok() && cohort_response.has_committed_response()) {
+      ++num_committed_responses;
+      response.mutable_committed_response()
+          ->mutable_response()
+          ->mutable_get_responses()
+          ->Add(cohort_response.committed_response().get_responses().begin(),
+                cohort_response.committed_response().get_responses().end());
+      metadata_by_transaction_[transaction_id]
+          .get_cohorts_already_responded.emplace(namespace_and_response.first);
+    }
+  }
+  if (num_committed_responses ==
+      metadata_by_transaction_[transaction_id].get_cohort_namespaces.size()) {
+    response.mutable_committed_response()->set_complete(true);
+    CleanUpTransactionMetadata(transaction_id);
+  }
+}
+
+void CoordinatorServer::CleanUpTransactionMetadata(
+    const std::string &transaction_id) {
+  response_by_transaction_[transaction_id].complete = true;
+  metadata_by_transaction_.erase(transaction_id);
+}
+
+grpc::Status CoordinatorServer::GetTransactionResult(
+    ServerContext *context, const GetTransactionResultRequest *request,
+    GetTransactionResultResponse *response) {
+  // If we already computed the response, return it.
+  if (response_by_transaction_[request->global_transaction_id()].complete) {
+    *response =
+        response_by_transaction_[request->global_transaction_id()].response;
+    return grpc::Status::OK;
+  }
+  // If we don't have metadata for the transaction, we can't ask the cohorts
+  // because we don't know who they are.
+  if (!metadata_by_transaction_.contains(request->global_transaction_id())) {
+    return grpc::Status(grpc::NOT_FOUND, "Could not find transaction");
+  }
+  TransactionMetadata &metadata =
+      metadata_by_transaction_[request->global_transaction_id()];
+  if (metadata.single_cohort_namespace.has_value()) {
```

Comment:
Could the multiple cohort code 'just work' for single cohort case?

Reply:
The only difference is that we don't invoke the blockchain for the single cohort so we always have to ask it what the response is (even if it's write only). If we change the code for the multiple cohorts case to ask all the cohorts for responses instead of just the ones with get operations, then it should 'just work'.

Reply:
I realized it still doesn't 'just work' because the single cohort needs to handle both commit and abort, while the multiple one can handle them separately since the blockchain told us which one was correct.